aten/src/ATen/core/type.cpp

```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
       "' but it contains an Any type. Any types cannot be members of modules, classes, or named tuples.");
 }
 
+namespace {
+bool isSpecialChar(char a) {
+  return a == '[' || a == ',' || a == ']';
+}
+
+class TypeParser {
+ public:
+  explicit TypeParser(const std::string& pythonStr)
+    : pythonStr_(pythonStr), start_(0) {}
+    
+  TypePtr parse() {
+    std::string token;
+    for (; start_ < pythonStr_.size() && !isSpecialChar(pythonStr_[start_]);
+           ++start_) {
+      if (pythonStr_[start_] != ' ') {
+        token += pythonStr_[start_];
+      }
+    }
+    if (token == "int") {
```

Comment:
@bhosmer - It would be nice to share the same parser logic but there are a couple issues. First, we want to serialize using Python type syntax, not our bespoke schema syntax. So we are actually not using the SchemaTypeParser syntax here, but the syntax used in our normal script parser. That parser parses types into an AST first, which is an extra bit of binary size we might not want in mobile.

Reply:
@zdevito aha! yeah, I think I haven't fully internalized that there are two type syntaxes :) gotcha re [ScriptTypeParser](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/script_type_parser.cpp) (as opposed to SchemaTypeParser) pulling in the AST machinery. (How much binary size do you think it would add?)

I guess I'm generically hoping there's some way to single-source some of the meaningful specifics, even if it's infeasible to just use the parser whole. Short of an infeasibly deep refactoring, maybe there are helpers we could use from both parsers, like https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/script_type_parser.cpp#L10? cc @iseeyuan

Reply:
@bhosmer Yes we could use the helpers. Thanks for bringing it up! But for the script_type_parser as a whole, it looks like it has dependencies on ir, tree, etc., where lite interpreter does not depend on. If we include those dependencies I'm pretty sure the size will be significantly bigger (around 1M) with no decoupling work.

Reply:
The basic string to type lookup table is shared now. For compose types (List, Tuple, Dict, etc.), a template function is used for single-element type. Couldn't find a quick and simple way to make them fully generic and reusable. I'm wondering if we could first make this version work so that we could unblock the operator work in general (which are blocking the automation flow in mobile), and revisit it when the operator work is done.

Reply:
@iseeyuan I think that's fine. If you wouldn't mind, could you make an issue for common lookup of generic types across these two parsers and assign it to me? I can follow up in parallel.

@zdevito qq on [this comment](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L641) - as far as I can tell this comment doesn't reflect current usage, right? Asking bc threading calls through a memo cache of instantiations would be one of the ways to factor out the common code here.

Reply:
Issue 31085 has been created and assigned to @bhosmer .